src/reachability/ai2.jl

```diff
@@ -18,25 +32,49 @@ Reachability analysis using split and join.
 Sound but not complete.
 
 # Reference
-T. Gehr, M. Mirman, D. Drashsler-Cohen, P. Tsankov, S. Chaudhuri, and M. Vechev,
+[1] T. Gehr, M. Mirman, D. Drashsler-Cohen, P. Tsankov, S. Chaudhuri, and M. Vechev,
 "Ai2: Safety and Robustness Certification of Neural Networks with Abstract Interpretation,"
 in *2018 IEEE Symposium on Security and Privacy (SP)*, 2018.
+
+## Note
+Efficient over-approximation of intersections and unions involving zonotopes relies on Theorem 3.1 of
+
+[2] Singh, G., Gehr, T., Mirman, M., Püschel, M., & Vechev, M. (2018). Fast
+and effective robustness certification. In Advances in Neural Information
+Processing Systems (pp. 10802-10813).
 """
-struct Ai2 <: Solver end
+struct Ai2{T<:Union{Hyperrectangle, Zonotope, HPolytope}} <: Solver end
+
+Ai2() = Ai2{Zonotope}()
+const Ai2h = Ai2{HPolytope}
+const Ai2z = Ai2{Zonotope}
+const Box = Ai2{Hyperrectangle}
 
 function solve(solver::Ai2, problem::Problem)
     reach = forward_network(solver, problem.network, problem.input)
     return check_inclusion(reach, problem.output)
 end
 
-forward_layer(solver::Ai2, layer::Layer, inputs::Vector{<:AbstractPolytope}) = forward_layer.(solver, layer, inputs)
+forward_layer(solver::Ai2, L::Layer, inputs::Vector) = forward_layer.(solver, L, inputs)
 
-function forward_layer(solver::Ai2, layer::Layer, input::AbstractPolytope)
-    outlinear = affine_map(layer, input)
-    relued_subsets = forward_partition(layer.activation, outlinear) # defined in ExactReach
+function forward_layer(solver::Ai2h, L::Layer, input::AbstractPolytope)
+    Ẑ = affine_map(L, input)
+    relued_subsets = forward_partition(L.activation, Ẑ) # defined in reachability.jl
     return convex_hull(relued_subsets)
 end
 
+# method for Zonotope and Hyperrectangle, if the input set isn't a Zonotope
+function forward_layer(solver::Ai2, L::Layer, input::AbstractPolytope)
+    X = overapproximate(input, Hyperrectangle)
+    return forward_layer(solver, L, X)
+end
+
+function forward_layer(solver::Ai2{T}, L::Layer, input::AbstractZonotope) where T<:Union{Hyperrectangle, Zonotope}
+    Ẑ = affine_map(L, input)
+    return overapproximate(Rectification(Ẑ), T)
```

Comment:
Note that `rectify(Ẑ)` is faster in the case of the Hyperrectangle, as `overapproximate(Rectification(Ẑ), Hyperrectangle)` does some extra computations that are not needed

Reply:
It's not that much faster, so it may not be worth the extra code, it's up to you
```julia
julia> H = rand(Hyperrectangle, dim=100);

julia> @btime rectify($H);
  4.720 ns (9 allocations: 5.31 KiB)

julia> @btime overapproximate(Rectification($H), Hyperrectangle);
  29.632 ns (10 allocations: 5.34 KiB)
```

Reply:
I don't mind another 1 line method definition 😄 
I'll add it in a moment.